src/lib/insights/client.ts

```diff
@@ -0,0 +1,128 @@
+import {
+  Hits,
+  InsightsClient,
+  InsightsClientMethod,
+  InsightsClientPayload,
+  InsightsClientWrapper,
+  Renderer,
+  RenderOptions,
+  Unmounter,
+  WidgetFactory,
+  Without,
+} from '../../types';
+import find from 'lodash/find';
+import uniq from 'lodash/uniq';
+
+const getSelectedHits = (hits: Hits<unknown>, selectedObjectIDs: string[]) => {
+  return selectedObjectIDs.map(objectID => {
+    const hit = find(hits, h => h.objectID === objectID);
+    if (typeof hit === 'undefined') {
+      throw new Error(
+        `Could not find objectID "${objectID}" passed to 'clickedObjectIDsAfterSearch' in the returned hits. This is necessary to infer absolute position and queryID.`
+      );
+    }
+    return hit;
+  });
+};
+
+const getQueryID = (selectedHits: Hits<unknown>) => {
+  const queryIDs = uniq(selectedHits.map(hit => hit.__queryID));
+  if (queryIDs.length > 1) {
+    throw new Error(
+      `Current API allows a single queryID. The objectIDs provided map to multiple queryIDs.`
+    );
+  }
+  const queryID = queryIDs[0];
+  if (typeof queryID !== 'string') {
+    throw new Error(
+      `Could not infer queryID. Ensure InstantSearch is configured with clickAnalytics: true `
```

Comment:
```suggestion
      'Could not infer query ID. Ensure `instantsearch` is configured with { clickAnalytics: true }'
```

Reply:
I skip this one. queryID is called queryID everywhere so it's best to keep it as it.
I kept the rest. see https://github.com/algolia/instantsearch.js/pull/3598/files/3aa8b4e978530c1784a3be8c7193d34a8a8dc1ed#r273981507